src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/Binding.java

```diff
@@ -0,0 +1,1029 @@
+ *
+ * int* f();
+ *
+ * Argument bindings:
+ * none
+ *
+ * Return bindings:
+ * 0: VM_LOAD(rax, long) // load a 'long' from the RAX register
+ *    BOX_ADDRESS // convert the 'long' into a 'MemoryAddress'
+ *
+ * --------------------
+ *
+ * typedef struct { // fits into single register
+ *   int x;
+ *   int y;
+ * } MyStruct;
+ *
+ * void f(MyStruct ms);
+ *
+ * Argument bindings:
+ * 0: BUFFER_LOAD(0, long.class) // From the struct's memory region, load a 'long' from offset '0'
+ *    VM_STORE(rcx, long.class) // and copy that into the RCX register
+ *
+ * Return bindings:
+ * none
+ *
+ * --------------------
+ *
+ * typedef struct { // does not fit into single register
+ *   long long x;
+ *   long long y;
+ * } MyStruct;
+ *
+ * void f(MyStruct ms);
+ *
+ * For the Windows ABI:
+ *
+ * Argument bindings:
+ * 0: COPY(16, 8) // copy the memory region containing the struct
+ *    BASE_ADDRESS // take the base address of the copy
+ *    UNBOX_ADDRESS // converts the base address to a 'long'
+ *    VM_STORE(rcx, long.class) // moves the 'long' into the RCX register
+ *
+ * Return bindings:
+ * none
+ *
+ * For the SysV ABI:
+ *
+ * Argument bindings:
+ * 0: DUP // duplicates the MemoryRegion operand
+ *    BUFFER_LOAD(0, long.class) // loads a 'long' from offset '0'
+ *    VM_STORE(rdx, long.class) // moves the long into the RDX register
+ *    BUFFER_LOAD(8, long.class) // loads a 'long' from offset '8'
+ *    VM_STORE(rcx, long.class) // moves the long into the RCX register
+ *
+ * Return bindings:
+ * none
+ *
+ * --------------------
+ *
+ * typedef struct { // fits into single register
+ *   int x;
+ *   int y;
+ * } MyStruct;
+ *
+ * MyStruct f();
+ *
+ * Argument bindings:
+ * none
+ *
+ * Return bindings:
+ * 0: ALLOCATE(GroupLayout(C_INT, C_INT)) // allocate a buffer with the memory layout of the struct
+ *    DUP // duplicate the allocated buffer
+ *    VM_LOAD(rax, long.class) // loads a 'long' from rax
+ *    BUFFER_STORE(0, long.class) // stores a 'long' at offset 0
+ *
+ * --------------------
+ *
+ * typedef struct { // does not fit into single register
+ *   long long x;
+ *   long long y;
+ * } MyStruct;
+ *
+ * MyStruct f();
+ *
+ * !! uses synthetic argument, which is a pointer to a pre-allocated buffer
+ *
+ * Argument bindings:
+ * 0: UNBOX_ADDRESS // unbox the MemoryAddress synthetic argument
+ *    VM_STORE(rcx, long.class) // moves the 'long' into the RCX register
+ *
+ * Return bindings:
+ * none
+ *
+ * --------------------
+ *
+ * void f(int dummy, ...); // varargs
+ *
+ * f(0, 10f); // passing a float
+ *
+ * Argument bindings:
+ * 0: VM_STORE(rcx, int.class) // moves the 'int dummy' into the RCX register
+ *
+ * 1: DUP // duplicates the '10f' argument
+ *    VM_STORE(rdx, float.class) // move one copy into the RDX register
+ *    VM_STORE(xmm1, float.class) // moves the other copy into the xmm2 register
+ *
+ * Return bindings:
+ * none
+ *
+ * --------------------
+ */
+public abstract class Binding {
```

Comment:
> Some design considerations, to consider later maybe.
> 
> The IR representation could be simplified to use record classes (which should be exiting preview in 16), implementing a Binding interface. The interpreter and specializer (compiler) could be separate if need be, operating on a sequence of instructions that just hold the data. Pattern matching could be used on the binding instances. It may be simpler and more efficient if the compiler generated explicit byte code rather than using MH combinators.

Good suggestions - I think this is good to consider post-integration.